File: scala/allergies/src/main/scala/Allergies.scala
object Allergen extends Enumeration {
  val Eggs, Peanuts, Shellfish, Strawberries, Tomatoes, Chocolate, Pollen, Cats = Value
}

object Allergies {
  def list(mask: Int) = Allergen.values.toList.filter(allergicTo(_, mask))

  def allergicTo(allergen: Allergen.Value, mask: Int) =  (1 << allergen.id & mask)  > 0

}

File: typescript/allergies/allergies.ts
export class Allergies {
  private MAP : {[key: string]: number} = {
    eggs: 1,
    peanuts: 2,
    shellfish: 4,
    strawberries: 8,
    tomatoes: 16,
    chocolate: 32,
    pollen: 64,
    cats: 128,
  }
  mask: number;

  constructor(allergenIndex: number) {
    this.mask = allergenIndex
  }

  list = () => Object.keys(this.MAP).filter(k => this.allergicTo(k));
  allergicTo = (allergen: string) => (this.MAP[allergen] & this.mask) > 0;
}
